### PR TITLE
Issue27 generate files

### PR DESCRIPTION
**Pull request by DannyPeck**
- Hatch Projects now old the relative path to their root.
- Hatch can now determine the operating system of the user.
- Hatch can now generate the necessary tup files in the correct locations, including the correct platform specific tup file.

### CODE REVIEW

**Comment by DannyPeck** [comment]
The hatch build command will now actually run the tup build command.

File: hatch/src/assets/builder.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    let contents = TestTupfile::new().to_string();
+
+    ProjectAsset::new(asset_path, TestTupfile::name(), contents)
+  }
+
+  pub fn tuprules(&self, project: &Project) -> ProjectAsset {
+    let project_path = project.path();
+    let project_kind = project.kind();
+    let contents = match *project_kind {
+      ProjectKind::Library(ref lib_type) =>  {
```
Comment (technetos) [inline hatch/src/assets/builder.rs]: Can this be done without { and } ?

File: hatch/src/assets/builder.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+pub struct Builder {
+  assets: Vec<ProjectAsset>,
+}
+
+impl Builder {
+  pub fn new() -> Builder {
+    Builder { assets: Vec::new() }
+  }
+
+  pub fn from(project: &Project) -> Builder {
```
Comment (technetos) [inline hatch/src/assets/builder.rs:21]: I really like this

File: hatch/src/assets/config.rs [hunk 3]
```diff
@@ -0,0 +1,36 @@
+use project::{ LibraryKind, ProjectKind };
+
+pub struct Config {
+  project: String,
+  lib_type: String,
+}
+
+impl Config {
+  pub fn new(name: &str, project_kind: &ProjectKind) -> Config {
+    Config { project: Config::project(name), lib_type: Config::lib_type(project_kind) }
```
Comment (technetos) [inline hatch/src/assets/config.rs:10]: Alright I prefer the readability of this but its also documented as the opposite in our style guide, one of them is going to change.

File: hatch/src/assets/config.rs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      ProjectKind::Library(LibraryKind::Shared) => "shared"
+    };
+
+    format!("LIB_TYPE = {}", kind)
+  }
+}
+
+impl ToString for Config {
+  fn to_string(&self) -> String {
+    [self.project.as_str(), self.lib_type.as_str()].join("\n")
```
Comment (technetos) [inline hatch/src/assets/config.rs:34]: This is great

File: hatch/src/assets/generator.rs [hunk 5]
```diff
@@ -0,0 +1,26 @@
+use assets::{ Asset, ProjectAsset };
+use hatch_error::{ HatchResult, ResultExt };
+use std::fs;
+use std::io::prelude::*;
+use std::io::Write;
+
+pub fn generate_one(asset: &ProjectAsset) -> HatchResult<()> {
+  let path = asset.path();
+  fs::create_dir_all(&path)?;
```
Comment (technetos) [inline hatch/src/assets/generator.rs]: Can we wrap these with better error messages?

```rust
fs::create_dir_all(&path).with_context(|e| {
  format!("Failed to create directory: `{}` : {}", &path, e)
})?;
```

File: hatch/src/assets/generator.rs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+  let file_path = path.join(asset.name());
+  let mut file = fs::File::create(&file_path)?;
+
+  let contents = asset.contents();
+  file.write_all(contents.as_bytes())?;
```
Comment (technetos) [inline hatch/src/assets/generator.rs]: See my comment about wrapping the ?'s with better error messages

File: hatch/src/assets/generator.rs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+  Ok(())
+}
+
+pub fn generate_all(assets: &Vec<ProjectAsset>) -> HatchResult<()> {
+  for asset in assets {
+    generate_one(asset)?;
```
Comment (technetos) [inline hatch/src/assets/generator.rs]: Can we make this:

```rust
generate_one(asset).with_context(|e| {
  format!("Failed to generate asset: `{}` : {}", asset.path(), e)
})?;
```

File: hatch/src/cli/commands/build.rs [hunk 8]
```diff
@@ -54,15 +36,32 @@ impl<'command> Command<'command> for Build {
     self.name
   }
 
-  fn execute(&self, args: &ArgMatches<'command>) -> Vec<HatchResult<Project>> {
-    let builder: Box<ProjectOps>;
+  fn execute(&self, args: &ArgMatches<'command>) -> HatchResult<Project> {
+    let project = task::read_project(self.project_path(args))?;
```
Comment (technetos) [inline hatch/src/cli/commands/build.rs]: See my comment about wrapping ?'s

**Review by technetos — CHANGES_REQUESTED** [review]

File: hatch/src/assets/builder.rs (hunk 1 above)
Comment (DannyPeck) [reply hatch/src/assets/builder.rs]: In theory, but the main reason I did it that way was due to line length.

File: hatch/src/assets/generator.rs (hunk 5 above)
Comment (DannyPeck) [reply hatch/src/assets/generator.rs]: Absolutely!

File: hatch/src/assets/config.rs (hunk 3 above)
Comment (DannyPeck) [reply hatch/src/assets/config.rs:10]: It does not explicitly state anything about struct construction :).

File: hatch/src/assets/config.rs (hunk 4 above)
Comment (technetos) [reply hatch/src/assets/config.rs:34]: actually do we want to use format! here?

File: hatch/src/assets/generator.rs [hunk 9]
```diff
@@ -6,20 +6,28 @@ use std::io::Write;
...
   let file_path = path.join(asset.name());
-  let mut file = fs::File::create(&file_path)?;
+  let mut file = fs::File::create(&file_path).with_context(|e| {
+    format!("Failed to create file: `{:?}` : {}", &file_path, e)
+  })?;
 
   let contents = asset.contents();
-  file.write_all(contents.as_bytes())?;
+  file.write_all(contents.as_bytes()).with_context(|e| {
+   format!("Failed to write contents to file: `{:?}` : {}", file_path, e)
```
Comment (technetos) [inline hatch/src/assets/generator.rs]:
```rust
format!("Failed to write contents to file: `{}` : {}", file_path.display(), e)
```

File: hatch/src/assets/generator.rs [hunk 10]
```diff
@@ -6,20 +6,28 @@ use std::io::Write;
...
+  })?;
 
   Ok(())
 }
 
 pub fn generate_all(assets: &Vec<ProjectAsset>) -> HatchResult<()> {
   for asset in assets {
-    generate_one(asset)?;
+    generate_one(asset).with_context(|e| {
+      format!("Failed to generate asset: `{:?}` : {}", asset.path(), e)
```
Comment (technetos) [inline hatch/src/assets/generator.rs]: see my comment about 
```rust 
.display 
```

File: hatch/src/cli/commands/build.rs [hunk 11]
```diff
@@ -37,8 +37,14 @@ impl<'command> Command<'command> for Build {
   }
 
   fn execute(&self, args: &ArgMatches<'command>) -> HatchResult<Project> {
-    let project = task::read_project(self.project_path(args))?;
-    task::generate_assets(&project)?;
+    let project_path = self.project_path(args);
+    let project = task::read_project(&project_path).with_context(|e| {
+      format!("Failed to read project `{:?}` : {}", project_path, e)
```
Comment (technetos) [inline hatch/src/cli/commands/build.rs]: why is this a
```rust
{:?}
```

**Review by technetos — CHANGES_REQUESTED** [review]

**Review by technetos — DISMISSED** [review]

File: .travis.yml [hunk 12]
```diff
@@ -1,10 +1,12 @@
-lanugage: rust
+language: rust
```
Comment (xmclark) [inline .travis.yml]: did you find a spellchecker for your text editor? hehe 🤣

**Review by xmclark — COMMENTED** [review]
I noticed a handful of files that do not have new lines at the end. Are we linuxy enough to care about that?

**Review by xmclark — DISMISSED** [review]
I'm mostly ok with this. Can you fix the new lines?

**Review by technetos — APPROVED** [review]

**Pull request merged**